### DISCUSSION TITLE
Feature request: Scoped _data folders that live amongst the templates

### DISCUSSION THREAD

**Discussion by kittee** [11ty/eleventy#4011] [Enhancement Queue]
Currently we can have template/directory specific data by naming it after the template or directory.

According to the docs, as of v0.5.4, this is the data hierarchy, using posts as an example (just copy/pasting it here for reference):

> 1. Content Template Front Matter Data
>     - merged with any Layout Front Matter Data
> 1. Template Data File (data is only applied to posts/subdir/my-first-blog-post.md)
>     - posts/subdir/my-first-blog-post.11tydata.js
>     - posts/subdir/my-first-blog-post.11tydata.json
>     - posts/subdir/my-first-blog-post.json
> 1. Directory Data File (data applies to all templates in posts/subdir/*)
>     - posts/subdir/subdir.11tydata.js
>     - posts/subdir/subdir.11tydata.json
>     - posts/subdir/subdir.json
> 1. Parent Directory Data File (data applies to all templates in posts/**/*, including subdirectories)
>     - posts/posts.11tydata.js
>     - posts/posts.11tydata.json
>     - posts/posts.json
> 1. Global Data Files in _data/* (available to all templates)

Using the posts example, I would like the option to have files named `posts/*.11tydata.js`, `posts/*.11tydata.json`, and/or `posts/*.json` and have the data in those files available to that template/directory.

Currently I'm using the global data to solve my problem since it allows wildcard matching with file names, but this isn't ideal because I don't need this data available in all my templates. Also I don't like the route of having one file named `posts/posts.json` for my case because I have a lot of data and I like keeping it in separate files for organizational purposes.

**Comment by zachleat** [comment]
Hm, I think I understand but can you walk me through an example of what you’re thinking?

So you have `posts/my-post-1.md` and `posts/my-post-2.md` and you want both of those to use a custom directory data file named something other than `posts/posts.json`?

Why not use `posts/my-post-1.json` and `posts/my-post-2.json` (or similar .11tydata.* files)?

**Comment by kittee** [comment]
So here's my situation:

I have a team page with details about team members. These team members don't have their own pages. So this is data that I want to use only on the team page, but I don't want to make a single `team.json` file because I like having a file for each team member.

So at first I was doing this:

```
|-- _data
|   |
|   |-- team
|       |
|       |-- jane-doe.json
|       |-- john-smith.json
|       |-- person-mcpersonface.json
|       |-- somebody-else.json
|
|-- team.liquid
```

After posting this issue and giving it more thought, I decided I hated this solution less:

```
|-- team
    |
    |-- _data
    |   |
    |   |-- jane-doe.js
    |   |-- john-smith.js
    |   |-- person-mcpersonface.js
    |   |-- somebody-else.js
    |
    |-- index.liquid
    |-- team.11tydata.js
```

And I'm using grunt-contrib-concat to turn the contents of `team/_data` into `team/team.11tydata.js`.

It working okay I guess? I still don't love it.

**Comment by edwardhorsford** [comment]
It sounds like you'd like to auto-import all `.js` and `.json` files in a folder and make that available to the templates in that folder? I'd quite like this too.

I'd be curious why `post.json` is supported but not `post.js`.

**Comment by jevets** [comment]
If it were me, I'd just drop em all into a `team.json` as an array of objects and be done with it. That said, I totally get your preferences.

I'd try a custom collection that your `team.liquid` (or `team/index.liquid`) could use, something like:

.eleventy.js:
```js
module.exports = function (config) {
  config.addCollection('team', function (collection) {
    return collection.getFilteredByGlob('team/_data/*.json')
  }

  return config
}
```

This also allows you to sort by some field, say w/ lodash `return _.sortBy(members, 'data.lastName')`

**Comment by kittee** [comment]
@jevets I _did_ try something like that actually, but `collection.getFilteredByGlob()` seems to only work on  template files.

**Comment by jevets** [comment]
@caramiki  Hmm, yeah that's right. 

You could add `.json` extension as a template extension, but that's probably not what you want. 
However, you could do that and use the new [`permalink: false`](https://www.11ty.io/docs/permalinks/#permalink%3A-false) feature on that whole directory... is an option, but still probably not what you want.

`<just-my-opinion>`

I'm not convinced this data can't be considered "global", though. Sure, it's available in all templates, even though you're only using it in one file — at least for now (and maybe forever). 

I often have a `_data/site.json` file that contains other global stuff, even though most of it is only used in a master layout or a chunk of it on a contact page. It's still "global" to me.

And I'd think the overhead of parsing and building `team.11tydata.js` via a separate task costs more than simply having a global data file which is read once during the build process. Probably faster build time to just consider it global in the first place.

`</just-my-opinion>`

- You're opposed to using front matter in the `team.liquid` file, right? (for whatever reason)

You could import each member's json file from a not-compiled `team.11tydata.js` and export all the objects in an array, instead of going through the whole grunt-contrib-concat. But this will grow cumbersome having to add a new `linda-faceymcfaceyperson.json` and add it to the `team.11tydata.js` file every time team members change.

Hopefully someone else can chime in with another idea or two. 
If it were me, I'd just consider it "global" data like your first structure and call it a day:

```
|-- _data
|   |
|   |-- team
|       |
|       |-- jane-doe.json
|       |-- john-smith.json
|       |-- person-mcpersonface.json
|       |-- somebody-else.json
|
|-- team.liquid
```

**Comment by kittee** [comment]
> However, you could do that and use the new [`permalink: false`](https://www.11ty.io/docs/permalinks/#permalink%3A-false) feature on that whole directory... is an option, but still probably not what you want.

Ooooh, I like this actually. Didn't know that `permalink: false` was a thing.

> I'm not convinced this data can't be considered "global", though. Sure, it's available in all templates, even though you're only using it in one file — at least for now (and maybe forever).

Yeah, when it comes down to it, all the data could be global, right? I just like small, well-organized files. ):

> * You're opposed to using front matter in the `team.liquid` file, right? (for whatever reason)

Yeah, it feels like a lot to dump in there.

**Comment by jevets** [comment]
@caramiki 

`permalink: false` is relatively new, added within the last release or two, I think. Been incredibly helpful for me, already.

Out of curiosity, have you settled on a route and what'd you do?

**Comment by kittee** [comment]
@jevets Well, I'm doing my last-mentioned solution currently (using grunt-contrib-concat to create a `team/team.11tydata.js` file from my multiple files) and that's been pushed to production. I'd like to go back, make each team member's file into a markdown file with front matter data, and use `permalink: false` instead but since it's already working and it's a low-priority project, refactoring is on the back burner.

We could close this issue if you want @zachleat

**Comment by zachleat** [comment]
Ah, I think I misunderstood what you were going for there @caramiki. I like the idea of scoped `_data` directories that only apply data to the directory and its children.

I’m going to move this into the enhancement queue.

**Comment by zachleat** [comment]
This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

The enhancement backlog can be found here: https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+

Don’t forget to upvote the top comment with 👍!

**Comment by zachleat** [comment]
Going to home base this one to #4077!
